File: scripts/check_md_links.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Find;
use File::Spec;

my $root = 'c:/ci4-local';
chdir $root or die "chdir $root: $!";

my @md;
find(sub { push @md, $File::Find::name if -f && /\.md$/i }, '.');

my $broken = 0;
my $trailing = 0;

foreach my $f (@md) {
    open my $fh, '<:encoding(UTF-8)', $f or next;
    my $line_no = 0;
    while (my $line = <$fh>) {
        $line_no++;
        if ($line =~ / $/) {
            print "TRAILING_WS: $f: line $line_no\n";
            $trailing++;
        }
        while ($line =~ /\[([^\]]+)\]\(([^)]+)\)/g) {
            my ($text, $link) = ($1, $2);
            next if $link =~ m{^https?://}i || $link =~ m{^mailto:}i || $link =~ /^#/;
            (my $clean = $link) =~ s/[#?].*$//;
            next if $clean eq '';
            my $target = File::Spec->rel2abs($clean, File::Spec->rel2abs($f));
            if (! -e $target) {
                print "BROKEN_LINK: $f -> [$text]($clean) (resolved: $target)\n";
                $broken++;
            }
        }
    }
    close $fh;
}

print "Checked ", scalar(@md), " markdown files. Trailing whitespace: $trailing. Broken links: $broken.\n";
exit($broken?2:0);

File: scripts/check_md_links.py
#!/usr/bin/env python3
import re
from pathlib import Path

root = Path('c:/ci4-local')
md_files = list(root.rglob('*.md'))
link_re = re.compile(r"\[([^\]]+)\]\(([^)]+)\)")

broken = []
trailing = []

for f in md_files:
    rel = f.relative_to(root)
    txt = f.read_text(encoding='utf-8')
    # trailing whitespace
    for i, line in enumerate(txt.splitlines(), start=1):
        if line.endswith(' '):
            trailing.append((str(rel), i))
    # links
    for m in link_re.finditer(txt):
        target = m.group(2).strip()
        if target.startswith('http') or target.startswith('mailto:') or target.startswith('#'):
            continue
        # remove optional title part: path "title"
        if ' ' in target and not target.startswith(' '):
            # handle URLs with spaces unlikely, but trim quotes
            t = target.split(' ')[0]
        else:
            t = target
        # remove fragment
        t = t.split('#')[0]
        # remove query
        t = t.split('?')[0]
        if t == '':
            continue
        target_path = (f.parent / t).resolve()
        if not target_path.exists():
            broken.append((str(rel), m.group(1), t))

print('Markdown files checked:', len(md_files))
if trailing:
    print('\nFiles with trailing whitespace:')
    for f,l in trailing:
        print(f' - {f}: line {l}')
else:
    print('\nNo trailing whitespace found.')

if broken:
    print('\nBroken relative links found:')
    for f, text, t in broken:
        print(f' - {f} -> [{text}]({t})')
    exit(2)
else:
    print('\nNo broken relative links found.')
    exit(0)
